File: web/js/screens/documents.js
/**
 * Màn hình: Tài liệu — nạp tệp và hỏi đáp có trích dẫn nguồn.
 *
 * Vào:  hồ sơ khách hàng đang chọn.
 * Ra:   danh sách tài liệu, khối hỏi đáp, lịch sử hỏi đáp.
 *
 * Ba thứ không được làm rơi so với bản Streamlit:
 *   A1.3  mỗi câu trả lời phải có trích dẫn trỏ đúng tệp và đúng đoạn, bấm mở được
 *   A1.6  câu hỏi suy luận tách rõ phần có trong tài liệu và phần suy đoán
 *   A1.7  nhãn `song ngữ song song` phải hiện, và đổi tay được nếu máy nhận sai
 */

import * as api from "../api.js";
import * as store from "../store.js";
import * as router from "../router.js";
import {
  el, pageHead, note, loading, empty, toast, ok, fail, withBusy, confirmDanger,
  fmtBytes, fmtDate,
} from "../ui.js";

const LANG_LABEL = {
  vi: ["badge-vi", "tiếng Việt"],
  en: ["badge-en", "tiếng Anh"],
  parallel: ["badge-ok", "song ngữ song song"],
  mixed: ["badge-mute", "trộn hai thứ tiếng"],
  unknown: ["badge-mute", "chưa rõ ngôn ngữ"],
};

const STATUS_LABEL = {
  ready: ["badge-ok", "sẵn sàng"],
  pending: ["badge-mute", "chờ xử lý"],
  extracting: ["badge-warn", "đang đọc"],
  chunking: ["badge-warn", "đang phân đoạn"],
  indexing: ["badge-warn", "đang lập chỉ mục"],
  error: ["badge-flag", "lỗi"],
};

function labelBadge(map, key) {
  const [cls, text] = map[key] ?? ["badge-mute", String(key)];
  return el(`span.badge.${cls}`, { text });
}

/* ============================== Nạp tệp ============================== */

function uploadBox(workspaceId, reload) {
  const input = el("input", {
    type: "file",
    multiple: true,
    accept: ".pdf,.docx,.doc,.txt,.md",
    style: "display:none",
    onchange: () => send(input.files),
  });

  const drop = el(
    "div.drop",
    {
      ondragover: (event) => {
        event.preventDefault();
        drop.dataset.over = "1";
      },
      ondragleave: () => {
        drop.dataset.over = "0";
      },
      ondrop: (event) => {
        event.preventDefault();
        drop.dataset.over = "0";
        send(event.dataTransfer.files);
      },
    },
    el("div.big", { text: "Kéo tệp vào đây, hoặc chọn từ máy" }),
    el("div.small", { text: "Nhận PDF · DOCX · DOC (Word cũ) · TXT · MD" }),
    el("button.btn.btn-primary", {
      type: "button",
      style: "margin-top:16px",
      text: "Chọn tệp",
      onclick: () => input.click(),
    }),
    input
  );

  async function send(fileList) {
    const files = Array.from(fileList ?? []);
    if (!files.length) return;

    const busy = el(
      "div.loading-row",
      null,
      el("span.spinner"),
      el("span", {
        text: `Đang nạp ${files.length} tệp — đọc, phân đoạn rồi lập chỉ mục. Tệp .doc cũ mất thêm chút thời gian vì phải mở qua Word.`,
      })
    );
    drop.replaceWith(busy);

    try {
      const result = await api.uploadDocuments(workspaceId, files);
      if (result.n_ready) ok(`Đã nạp xong ${result.n_ready}/${result.n_total} tệp.`);
      // Một tệp lỗi KHÔNG làm hỏng cả lô (A1.9) — báo riêng từng tệp, nêu lý do cụ thể
      for (const item of result.results ?? []) {
        if (item.status !== "ready") {
          fail(`${item.filename}: ${item.error_message ?? "không nạp được"}`);
        }
      }
      await store.touchWorkspace(workspaceId);
      reload();
    } catch (error) {
      busy.replaceWith(drop);
      fail(error.message);
    }
  }

  return drop;
}

/* ============================== Danh sách tài liệu ============================== */

function showModal(...content) {
  const backdrop = el("div.modal-backdrop", {
    onclick: (event) => {
      if (event.target === backdrop) backdrop.remove();
    },
  });
  const close = () => backdrop.remove();
  backdrop.append(
    el(
      "div.modal",
      { role: "dialog", "aria-modal": "true" },
      ...content,
      el("div.modal-foot", null, el("button.btn", { type: "button", text: "Đóng", onclick: close }))
    )
  );
  document.body.append(backdrop);
  backdrop.querySelector("button")?.focus();
  const onKey = (event) => {
    if (event.key === "Escape") {
      close();
      document.removeEventListener("keydown", onKey);
    }
  };
  document.addEventListener("keydown", onKey);
}

function documentRow(doc, reload) {
  const actions = el(
    "div.btn-row",
    null,
    el("button.btn.btn-sm", {
      type: "button",
      text: "Xem trước",
      onclick: async (event) => {
        await withBusy(event.currentTarget, "…", async () => {
          const preview = await api.previewDocument(doc.id);
          showModal(
            el("h2", { text: doc.filename }),
            el("p.card-sub", {
              text:
                `${preview.showing}/${preview.total_chunks} đoạn · ` +
                `${doc.n_chars} ký tự · đọc bằng ${doc.extractor ?? "?"}`,
            }),
            doc.extraction_quality === "low" &&
              note(
                "warn",
                "Trích xuất chất lượng thấp — tệp .doc cũ đọc bằng lớp dự phòng. " +
                  "Soát kỹ nội dung dưới đây trước khi tin vào câu trả lời."
              ),
            el(
              "div.modal-body",
              null,
              (preview.chunks ?? []).map((chunk) =>
                el(
                  "div",
                  { style: "margin-bottom:14px" },
                  el("div", {
                    style: "font-family:var(--font-mono);font-size:var(--t-micro);color:var(--ink-4)",
                    text: `${chunk.locator ?? `đoạn ${chunk.chunk_index}`} · ${chunk.lang ?? "?"}`,
                  }),
                  el("div", { text: chunk.text })
                )
              )
            )
          );
        });
      },
    }),
    // Máy nhận sai ngôn ngữ thì phải sửa tay được; đổi nhãn thì trích thuật ngữ lại
    el(
      "select.select",
      {
        style: "width:auto;padding:.25rem .5rem;font-size:var(--t-xs)",
        "aria-label": `Ngôn ngữ của ${doc.filename}`,
        onchange: async (event) => {
          try {
            await api.setDocumentLanguage(doc.id, event.target.value);
            ok("Đã đổi nhãn ngôn ngữ. Lần trích thuật ngữ sau sẽ dùng nhãn mới.");
            reload();
          } catch (error) {
            fail(error.message);
          }
        },
      },
      ["vi", "en", "parallel", "mixed", "unknown"].map((value) =>
        el("option", { value, text: LANG_LABEL[value][1], selected: value === doc.language })
      )
    ),
    el("button.btn.btn-sm.btn-danger", {
      type: "button",
      text: "Xoá",
      onclick: async () => {
        const yes = await confirmDanger({
          title: `Xoá “${doc.filename}”?`,
          body:
            "Tệp và toàn bộ chỉ mục của nó sẽ bị xoá vĩnh viễn. Nội dung này sẽ không còn " +
            "xuất hiện trong bất kỳ câu trả lời nào sau đó.",
          confirmLabel: "Xoá tài liệu",
        });
        if (!yes) return;
        try {
          await api.deleteDocument(doc.id);
          ok("Đã xoá tài liệu.");
          await store.touchWorkspace();
          reload();
        } catch (error) {
          fail(error.message);
        }
      },
    })
  );

  return el(
    "div.doc-row",
    null,
    el(
      "div.grow",
      null,
      el("div.name", { text: doc.filename }),
      el(
        "div",
        { style: "display:flex;gap:6px;align-items:center;flex-wrap:wrap;margin-top:5px" },
        labelBadge(STATUS_LABEL, doc.status),
        labelBadge(LANG_LABEL, doc.language),
        doc.language_source === "manual" && el("span.badge.badge-mute", { text: "bạn đã đổi nhãn" }),
        doc.extraction_quality === "low" &&
          el("span.badge.badge-warn", { text: "trích xuất chất lượng thấp" }),
        el("span.meta", {
          text: `${String(doc.ext).toUpperCase()} · ${fmtBytes(doc.size_bytes)} · ${doc.n_chunks} đoạn`,
        })
      ),
      doc.error_message && el("div", { style: "margin-top:8px" }, note("error", doc.error_message))
    ),
    actions
  );
}

/* ============================== Hỏi đáp ============================== */

function citationChip(citation) {
  return el("button.citation", {
    type: "button",
    text: `${citation.filename} · ${citation.locator ?? ""}`.trim(),
    title: "Bấm để xem nguyên văn đoạn được trích",
    onclick: () =>
      showModal(
        el("h2", { text: citation.filename }),
        el("p.card-sub", { text: citation.locator ?? "" }),
        el("div.modal-body", null, el("div", { style: "white-space:pre-wrap", text: citation.snippet ?? "" }))
      ),
  });
}

const CONFIDENCE_NOTE = {
  low: "Câu trả lời dựa trên ngữ cảnh hạn chế — nên kiểm tra lại nguồn trước khi dùng.",
  medium: null,
  high: null,
};

function answerBlock(result) {
  const box = el("div.card", { style: "margin-top:8px" });

  // Không tìm thấy thì nói không tìm thấy, và mở đường sang tìm web (A1.4)
  if (result.found === false) {
    box.append(
      note("warn", result.answer),
      el("button.btn", {
        type: "button",
        style: "margin-top:8px",
        text: "Tìm trên web thay vì trong tài liệu",
        onclick: () => router.go("research"),
      })
    );
    return box;
  }

  const confNote = CONFIDENCE_NOTE[result.confidence];
  if (confNote) box.append(note("warn", confNote));
  for (const w of result.warnings ?? []) box.append(note("warn", w));

  box.append(el("div", { style: "white-space:pre-wrap", text: result.answer }));

  // Số liệu là thứ phiên dịch sai là hỏng — tách riêng cho dễ soát
  if ((result.key_figures ?? []).length) {
    box.append(
      el(
        "div",
        { style: "margin-top:12px;display:flex;gap:6px;flex-wrap:wrap;align-items:center" },
        el("span", { style: "font-size:var(--t-xs);color:var(--ink-3)", text: "Số liệu:" }),
        result.key_figures.map((figure) => el("span.badge.badge-vi", { text: figure }))
      )
    );
  }

  // Phần suy luận tách hẳn khỏi phần đọc được từ tài liệu (A1.6)
  if (result.inference) {
    box.append(
      el(
        "div",
        {
          style:
            "margin-top:14px;padding-top:12px;border-top:1px dashed var(--line-2)",
        },
        el(
          "div",
          { style: "margin-bottom:6px" },
          el("span.badge.badge-warn", { text: "suy luận — không đọc được trực tiếp từ tài liệu" })
        ),
        el("div.card-sub", { style: "white-space:pre-wrap", text: result.inference })
      )
    );
  }

  const citations = result.citations ?? [];
  if (citations.length) {
    box.append(
      el(
        "div",
        { style: "margin-top:14px;padding-top:12px;border-top:1px solid var(--line)" },
        el("div", {
          style: "font-size:var(--t-xs);color:var(--ink-3);margin-bottom:6px",
          text: `${citations.length} trích dẫn — bấm để xem nguyên văn`,
        }),
        el("div", null, citations.map(citationChip))
      )
    );
  } else {
    box.append(
      el(
        "div",
        { style: "margin-top:12px" },
        note("warn", "Câu trả lời này không có trích dẫn — hãy tự kiểm lại trong tài liệu.")
      )
    );
  }
  return box;
}

const SAMPLE_QUESTIONS = [
  "Tóm tắt thông tin về khách hàng",
  "Cho tôi thông tin về dự án",
  "Dự đoán nội dung sẽ trao đổi trong buổi nghiệm thu",
];

function askBlock(workspaceId, documents) {
  const input = el("textarea.textarea", {
    rows: 2,
    placeholder: "Hỏi bằng tiếng Việt hoặc tiếng Anh…",
  });
  const scope = el(
    "select.select",
    { style: "width:auto", "aria-label": "Phạm vi hỏi" },
    el("option", { value: "", text: "Toàn bộ hồ sơ" }),
    documents.filter((d) => d.status === "ready").map((d) =>
      el("option", { value: String(d.id), text: d.filename })
    )
  );
  const submit = el("button.btn.btn-primary", { type: "submit", text: "Hỏi" });
  const answers = el("div");

  async function ask(question) {
    input.value = question;
    await withBusy(submit, "Đang tra…", async () => {
      const result = await api.askDocuments({
        workspace_id: workspaceId,
        question,
        document_ids: scope.value ? [Number(scope.value)] : null,
      });
      answers.prepend(
        el(
          "div",
          { style: "margin-top:16px" },
          el("div", { style: "font-weight:600", text: question }),
          answerBlock(result)
        )
      );
      input.value = "";
    });
  }

  const form = el(
    "form.card",
    {
      onsubmit: (event) => {
        event.preventDefault();
        const question = input.value.trim();
        if (!question) {
          toast("warn", "Chưa nhập câu hỏi.");
          return;
        }
        ask(question);
      },
    },
    el("div.card-title", { text: "Hỏi trên tài liệu" }),
    el("p.card-sub", {
      style: "margin-bottom:12px",
      text:
        "Câu trả lời luôn kèm trích dẫn trỏ về đúng tệp và đúng đoạn. " +
        "Không tìm thấy thì nói không tìm thấy, không suy diễn thành câu khẳng định.",
    }),
    input,
    el("div.btn-row", { style: "margin-top:12px" }, scope, submit),
    el(
      "div.btn-row",
      { style: "margin-top:12px" },
      el("span", { style: "font-size:var(--t-xs);color:var(--ink-3)", text: "Hỏi nhanh:" }),
      SAMPLE_QUESTIONS.map((q) =>
        el("button.btn.btn-sm.btn-ghost", { type: "button", text: q, onclick: () => ask(q) })
      )
    )
  );

  return el("div", null, form, answers);
}

/* ============================== Render ============================== */

export async function render(root) {
  const current = store.currentWorkspace();
  if (!current) {
    root.append(
      pageHead("CHUẨN BỊ", "Tài liệu", "Nạp tài liệu và hỏi đáp có trích dẫn nguồn"),
      empty("Chưa chọn hồ sơ khách hàng", "Tạo hoặc chọn một hồ sơ ở Bảng điều khiển trước.")
    );
    return;
  }

  root.append(
    pageHead(
      "CHUẨN BỊ",
      "Tài liệu",
      "Nạp tài liệu của buổi làm việc, rồi hỏi bằng tiếng Việt — câu trả lời luôn chỉ ra nguồn"
    )
  );

  const slot = el("div");
  root.append(slot);

  async function reload() {
    slot.innerHTML = "";
    slot.append(loading("Đang tải danh sách tài liệu…"));

    let documents;
    try {
      documents = await api.listDocuments(current.id);
    } catch (error) {
      slot.innerHTML = "";
      slot.append(note("error", error.message));
      return;
    }

    slot.innerHTML = "";
    slot.append(uploadBox(current.id, reload));

    if (documents.length) {
      slot.append(
        el("div.card", { style: "margin-top:16px;padding:0" }, documents.map((doc) => documentRow(doc, reload)))
      );

      const parallel = documents.filter((d) => d.language === "parallel");
      if (parallel.length) {
        slot.append(
          el(
            "div",
            { style: "margin-top:12px" },
            note(
              "ok",
              `${parallel.length} tài liệu song ngữ song song — nguồn thuật ngữ đáng tin nhất, vì cặp dịch ` +
                "do người thật làm. Cũng dùng làm bản dịch chuẩn ⭐⭐⭐ khi chấm điểm."
            )
          )
        );
      }
    } else {
      slot.append(
        el(
          "div",
          { style: "margin-top:16px" },
          note("warn", "Chưa có tài liệu nào. Nạp hợp đồng, biên bản hoặc bài phát biểu của buổi làm việc.")
        )
      );
    }

    if (documents.some((d) => d.status === "ready")) {
      slot.append(el("div.section", null, askBlock(current.id, documents)));

      try {
        const items = await api.qaHistory(current.id);
        if (items.length) {
          slot.append(
            el(
              "div.section",
              null,
              el("h2", { text: "Đã hỏi trước đó" }),
              el(
                "div.card",
                { style: "padding:0" },
                items.slice(0, 12).map((item) =>
                  el(
                    "div",
                    { style: "padding:12px 20px;border-bottom:1px solid var(--line)" },
                    el("div", { style: "font-weight:600;font-size:var(--t-sm)", text: item.question }),
                    el("div.card-sub", {
                      style: "margin-top:4px",
                      text:
                        String(item.answer ?? "").slice(0, 220) +
                        (String(item.answer ?? "").length > 220 ? "…" : ""),
                    }),
                    el("div", {
                      style:
                        "font-family:var(--font-mono);font-size:var(--t-micro);color:var(--ink-4);margin-top:6px",
                      text: fmtDate(item.created_at),
                    })
                  )
                )
              )
            )
          );
        }
      } catch {
        /* lịch sử hỏng không đáng chặn màn hình chính */
      }
    }
  }

  await reload();
}

File: backend/routes/documents.py
"""Routes Module 1: nạp tài liệu, xem trước, xoá, và hỏi đáp có trích dẫn.

Input:  tệp tải lên (multipart) hoặc câu hỏi.
Output: trạng thái từng tệp, hoặc câu trả lời kèm trích dẫn.

Một tệp lỗi không làm hỏng cả lô (spec A1.9): mỗi tệp trả về trạng thái riêng.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from fastapi import APIRouter, File, Form, HTTPException, Query, UploadFile

from backend.config import settings
from backend.db import get_conn
from backend.rag import ingest, qa
from backend.rag.extractors import SUPPORTED_EXTENSIONS
from backend.rag.language import classify

router = APIRouter(prefix="/documents", tags=["documents"])


def _require_workspace(workspace_id: int) -> None:
    with get_conn() as conn:
        row = conn.execute("SELECT 1 FROM workspaces WHERE id = ?", (workspace_id,)).fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="Không tìm thấy hồ sơ khách hàng này.")


@router.get("")
def list_documents(workspace_id: int = Query(...)) -> list[dict[str, Any]]:
    _require_workspace(workspace_id)
    with get_conn() as conn:
        rows = conn.execute(
            """
            SELECT id, filename, ext, size_bytes, language, language_source,
                   extraction_quality, extractor, status, error_message,
                   n_chars, n_chunks, created_at
            FROM documents WHERE workspace_id = ?
            ORDER BY created_at DESC, id DESC
            """,
            (workspace_id,),
        ).fetchall()
    return [dict(row) for row in rows]


@router.post("/upload")
async def upload_documents(
    workspace_id: int = Form(...),
    files: list[UploadFile] = File(...),
) -> dict[str, Any]:
    """Nạp một hoặc nhiều tệp. Trả về kết quả riêng cho từng tệp."""
    _require_workspace(workspace_id)

    results: list[dict[str, Any]] = []
    max_bytes = settings.max_upload_mb * 1024 * 1024

    for upload in files:
        name = upload.filename or "tai-lieu"
        ext = Path(name).suffix.lower()

        # --- Chặn sớm: định dạng không hỗ trợ ---
        if ext not in SUPPORTED_EXTENSIONS:
            results.append(
                {
                    "filename": name,
                    "status": "error",
                    "error": (
                        f"Chưa hỗ trợ định dạng '{ext}'. Các định dạng nhận được: "
                        f"{', '.join(sorted(SUPPORTED_EXTENSIONS))}."
                    ),
                }
            )
            continue

        data = await upload.read()

        if not data:
            results.append(
                {"filename": name, "status": "error", "error": f"Tệp '{name}' rỗng (0 byte)."}
            )
            continue

        if len(data) > max_bytes:
            results.append(
                {
                    "filename": name,
                    "status": "error",
                    "error": (
                        f"Tệp '{name}' nặng {len(data) / 1024 / 1024:.1f} MB, vượt ngưỡng "
                        f"{settings.max_upload_mb} MB. Hãy tách nhỏ tệp rồi nạp lại."
                    ),
                }
            )
            continue

        stored_path = ingest.save_upload(workspace_id, name, data)

        # --- Trùng nội dung: hỏi thay vì âm thầm nhân đôi (spec A1.10) ---
        duplicate = ingest.find_duplicate(workspace_id, ingest._content_hash(stored_path))
        if duplicate:
            stored_path.unlink(missing_ok=True)
            results.append(
                {
                    "filename": name,
                    "status": "duplicate",
                    "error": (
                        f"Nội dung tệp này trùng với '{duplicate['filename']}' đã nạp trước đó. "
                        "Bỏ qua để tránh câu trả lời bị lặp. Muốn thay thế thì xoá tệp cũ trước."
                    ),
                }
            )
            continue

        result = ingest.ingest_file(workspace_id, stored_path, name)
        results.append(
            {
                "document_id": result.document_id,
                "filename": result.filename,
                "status": result.status,
                "language": result.language,
                "language_label": result.language_label,
                "extractor": result.extractor,
                "quality": result.quality,
                "n_chars": result.n_chars,
                "n_chunks": result.n_chunks,
                "warnings": result.warnings,
                "error": result.error,
            }
        )

    n_ok = sum(1 for r in results if r.get("status") == "ready")
    return {
        "results": results,
        "n_total": len(results),
        "n_ready": n_ok,
        "n_failed": len(results) - n_ok,
    }


@router.get("/{document_id}/preview")
def preview_document(document_id: int, max_chunks: int = Query(default=12, ge=1, le=200)) -> dict[str, Any]:
    """Xem nội dung đã trích xuất — để chuyên gia tự đánh giá chất lượng trích xuất."""
    with get_conn() as conn:
        doc = conn.execute("SELECT * FROM documents WHERE id = ?", (document_id,)).fetchone()
        if doc is None:
            raise HTTPException(status_code=404, detail="Không tìm thấy tài liệu này.")
        chunks = conn.execute(
            """
            SELECT id, chunk_index, locator, lang, text FROM document_chunks
            WHERE document_id = ? ORDER BY chunk_index LIMIT ?
            """,
            (document_id, max_chunks),
        ).fetchall()

    return {
        "document": dict(doc),
        "chunks": [dict(c) for c in chunks],
        "showing": len(chunks),
        "total_chunks": int(doc["n_chunks"]),
    }


@router.patch("/{document_id}/language")
def set_language(document_id: int, language: str = Query(...)) -> dict[str, Any]:
    """Đổi nhãn ngôn ngữ thủ công khi máy nhận sai (edge case §4.1 của spec)."""
    allowed = {"vi", "en", "parallel", "mixed", "unknown"}
    if language not in allowed:
        raise HTTPException(
            status_code=400, detail=f"Nhãn ngôn ngữ phải là một trong: {', '.join(sorted(allowed))}."
        )
    with get_conn() as conn:
        cursor = conn.execute(
            "UPDATE documents SET language = ?, language_source = 'manual' WHERE id = ?",
            (language, document_id),
        )
        if cursor.rowcount == 0:
            raise HTTPException(status_code=404, detail="Không tìm thấy tài liệu này.")
    return {"document_id": document_id, "language": language, "language_source": "manual"}


@router.post("/{document_id}/reindex")
def reindex_document(document_id: int) -> dict[str, Any]:
    """Nạp lại một tệp đã có (dùng khi lần trước lỗi, hoặc sau khi đổi nhãn ngôn ngữ)."""
    with get_conn() as conn:
        doc = conn.execute("SELECT * FROM documents WHERE id = ?", (document_id,)).fetchone()
    if doc is None:
        raise HTTPException(status_code=404, detail="Không tìm thấy tài liệu này.")

    path = Path(str(doc["stored_path"]))
    if not path.exists():
        raise HTTPException(
            status_code=410,
            detail=f"Tệp gốc '{doc['filename']}' không còn trên đĩa. Hãy tải lên lại.",
        )

    workspace_id = int(doc["workspace_id"])
    ingest.delete_document(document_id)
    result = ingest.ingest_file(workspace_id, path, str(doc["filename"]))
    return {
        "document_id": result.document_id,
        "status": result.status,
        "language": result.language,
        "n_chunks": result.n_chunks,
        "error": result.error,
        "warnings": result.warnings,
    }


@router.delete("/{document_id}")
def delete_document(document_id: int) -> dict[str, Any]:
    if not ingest.delete_document(document_id):
        raise HTTPException(status_code=404, detail="Không tìm thấy tài liệu này.")
    return {"deleted": document_id, "message": "Đã xoá tài liệu và toàn bộ chỉ mục của nó."}


@router.post("/ask")
def ask_documents(payload: dict[str, Any]) -> dict[str, Any]:
    """Hỏi đáp trên tài liệu. Câu trả lời luôn kèm trích dẫn hoặc nói rõ không tìm thấy."""
    workspace_id = payload.get("workspace_id")
    question = str(payload.get("question", "")).strip()
    document_ids = payload.get("document_ids") or None

    if not workspace_id:
        raise HTTPException(status_code=400, detail="Thiếu hồ sơ khách hàng.")
    if not question:
        raise HTTPException(status_code=400, detail="Câu hỏi đang để trống.")

    _require_workspace(int(workspace_id))

    answer = qa.ask(
        workspace_id=int(workspace_id), question=question, document_ids=document_ids
    )
    return {
        "question": answer.question,
        "answer": answer.answer,
        "found": answer.found,
        "confidence": answer.confidence,
        "inference": answer.inference,
        "key_figures": answer.key_figures,
        "warnings": answer.warnings,
        "citations": [
            {
                "document_id": c.document_id,
                "filename": c.filename,
                "locator": c.locator,
                "snippet": c.snippet,
                "chunk_id": c.chunk_id,
            }
            for c in answer.citations
        ],
    }


@router.get("/qa-history")
def qa_history(workspace_id: int = Query(...), limit: int = Query(default=50, ge=1, le=500)) -> list[dict[str, Any]]:
    _require_workspace(workspace_id)
    return qa.history(workspace_id, limit=limit)
